modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/Implementor.java

```diff
@@ -38,151 +32,104 @@
+        return new ProjectNode(ctx, visit(rel.getInput()), projection);
     }
 
     /** {@inheritDoc} */
     @Override public Node<Object[]> visit(IgniteJoin rel) {
-        assert !stack.isEmpty();
-
-        JoinNode res = new JoinNode(stack.pop(), factory.joinExpression(root, rel.getCondition(), rel.getLeft().getRowType(), rel.getRight().getRowType()));
-
-        stack.push(res.sink(1));
-        stack.push(res.sink(0));
-
-        res.sources(sources(rel.getInputs()));
-
-        return res;
+        BiFunction<Object[], Object[], Object[]> expression = factory.joinExpression(ctx, rel.getCondition(), rel.getLeft().getRowType(), rel.getRight().getRowType());
+        return new JoinNode(ctx, visit(rel.getLeft()), visit(rel.getRight()), expression);
     }
 
     /** {@inheritDoc} */
     @Override public Node<Object[]> visit(IgniteTableScan rel) {
-        assert !stack.isEmpty();
-
-        Iterable<Object[]> source = rel.getTable().unwrap(ScannableTable.class).scan(root);
-
-        return new ScanNode(stack.pop(), source);
+        Iterable<Object[]> source = rel.getTable().unwrap(ScannableTable.class).scan(ctx);
+        return new ScanNode(ctx, source);
     }
 
     /** {@inheritDoc} */
     @Override public Node<Object[]> visit(IgniteReceiver rel) {
-        throw new AssertionError(); // TODO
-    }
+        RelSource source = rel.source();
+        MailboxRegistry registry = ctx.mailboxRegistry();
 
-    /** {@inheritDoc} */
-    @Override public Node<Object[]> visit(IgniteExchange rel) {
-        throw new AssertionError();
+        // Corresponding outbox fragment ID is used as exchange ID as well.
+        Inbox<Object[]> inbox = (Inbox<Object[]>) registry.register(new Inbox<>(ctx, source.fragmentId(), source.fragmentId()));
+
+        // here may be an already created (to consume rows from remote nodes) inbox
+        // without proper context, we need to init it with a right one.
+        inbox.init(ctx, source.mapping().nodes(), factory.comparator(ctx, rel.collations(), rel.getRowType()));
+
+        return inbox;
     }
 
     /** {@inheritDoc} */
     @Override public Node<Object[]> visit(IgniteRel rel) {
-        throw new AssertionError();
+        return rel.accept(this);
     }
 
-    /** */
-    private Source source(RelNode rel) {
-        if (rel.getConvention() != IgniteConvention.INSTANCE)
-            throw new IllegalStateException("INTERPRETABLE is required.");
-
-        return ((IgniteRel) rel).accept(this);
+    /** {@inheritDoc} */
+    @Override public Node<Object[]> visit(IgniteExchange rel) {
+        throw new AssertionError();
     }
 
     /** */
-    private List<Source> sources(List<RelNode> rels) {
-        ArrayList<Source> res = new ArrayList<>(rels.size());
-
-        for (RelNode rel : rels) {
-            res.add(source(rel));
-        }
-
-        return res;
+    private Node<Object[]> visit(RelNode rel) {
+        return visit(igniteRel(rel));
```

Comment:
It's used just to cheek convention and cast an abstract rel node to specific one. Since a relation graph operates with abstract nodes the cast is necessary (to use ignite specific apply() methods) probably it's good to make an overloaded method instead (which accepts an abstract rel node, casts it and invokes igniteRel.accept() method)

Reply:
It's a kind of assertion. I think it's better not to mix it with program logic and use something like this:
```
RelNode rel = ...;
if (U.assertionsEnabled()) {
  checkIginteConvention(rel); // Entire query tree is checked here.
}
IgniteRel igniteRel = (IgniteRel)rel;
new Visitor().go(igniteRel)
```

Reply:
It's ok to do conditional check (if assertions enabled), but the method is really useful, because such logic is needed in several places.

Reply:
It is not a logic, it is just an assertion which pollutes the code. I am not sure it is useful.

Reply:
Ok, I've reviewed all places, since in all cases the method is used a different type than IgniteRel signals there is a bug in code, it's OK to just cast a rel to IgniteRel without any checking. Removed the method.